src/cdk-experimental/menu/menu-aim.ts

```diff
@@ -0,0 +1,271 @@
+ * coordinates.
+ */
+type SubmenuPoints = {bottom: number; top: number; x: number; midX: number; endX: number};
+
+/** Whether the given line defined by the slope and y intercept falls within the submenu. */
+function isWithinSubmenu(submenuPoints: SubmenuPoints, _slope: number, yintercept: number) {
+  return (
+    (_slope * submenuPoints.x + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.x + yintercept < submenuPoints.bottom) ||
+    (_slope * submenuPoints.midX + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.midX + yintercept < submenuPoints.bottom) ||
+    (_slope * submenuPoints.endX + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.endX + yintercept < submenuPoints.bottom)
+  );
+}
+/**
+ * TargetMenuAim attempts to predict if a user is moving into a submenu. It calculates the
+ * trajectory of the users mouse movement in the current menu in order to determine if the
+ * mouse is moving towards an opened submenu.
+ *
+ * The determination is made by calculating the slope of the users last NUM_POINTS moves where each
+ * pair of points determines if the trajectory line points into the submenu. It uses consensus
+ * approach by checking if at least NUM_POINTS / 2 pairs determine that the user is moving towards
+ * to submenu.
+ */
+@Injectable()
+export class TargetMenuAim implements MenuAim, OnDestroy {
+  /** The last NUM_POINTS mouse move events. */
+  private readonly _points: Point[] = [];
+
+  /** Reference to the root menu in which we are tracking mouse moves. */
+  private _menu: Menu;
+
+  /** Reference to the root menus mouse manager. */
+  private _pointerTracker: PointerFocusTracker<Toggler & FocusableElement>;
+
+  /** The id associated with the current timeout call waiting to resolve. */
+  private _timeoutId: number | null;
+
+  /** Emits when this service is destroyed. */
+  private readonly _destroyed: Subject<void> = new Subject();
+
+  constructor(private readonly _ngZone: NgZone) {}
+
+  /** Set the mouse manager tracking the menu's menu items. */
+  withMouseManager(pointerTracker: PointerFocusTracker<FocusableElement & Toggler>): this {
+    this._pointerTracker = pointerTracker;
+    return this;
+  }
+
+  /** Set the menu for which this tracks mouse movements in. */
+  withMenu(menu: Menu): this {
+    this._menu = menu;
+    this._subscribeToMouseMoves();
+    return this;
+  }
+
+  /**
+   * Calls the `doToggle` callback when it is deemed that the user is not moving towards
+   * the submenu.
+   * @param doToggle the function called when the user is not moving towards the submenu.
+   */
+  toggle(doToggle: () => void) {
+    // If the menu is horizontal the sub-menus open below and there is no risk of premature
+    // closing of any sub-menus therefore we automatically resolve the callback.
+    if (this._menu.orientation === 'horizontal') {
+      doToggle();
+    }
```

Comment:
I don't necessarily know that this is true; a horizontal menu could still technically have menus popping up in overlays anywhere

Reply:
By default horizontal menus open [below or above](https://github.com/angular/components/blob/master/src/cdk-experimental/menu/menu-item-trigger.ts#L283-L286). 

The reason for ignoring horizontal menus is for example in a regular menu bar (Google Docs) they are next to each other, so if you open the file menu there's no chance of you hovering over the edit trigger when going into the the file menu.

Reply:
For the default set up it makes sense, but if someone makes a custom horizontal menu it might come up. I think it's fine to leave it this way for now, though, and see if a real case arises.

Reply:
Ah, didn't consider that. I'll leave for now since its easy enough to remove